### PR TITLE
Add facility to support with-contexts

### PR DESCRIPTION
**Pull request by sklam**
Adds pluggable facility to support `with` contexts.  Two builtin contextmanagers are implemented for testing.

### CODE REVIEW

File: numba/withcontexts.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        _clear_blocks(blocks, body_blocks)
+        return dispatcher
+
+
+call_context = _CallContextType()
+
+
+def _bypass_with_context(blocks, blk_start, blk_end, forwardvars):
+    """Given the starting and ending block of the with-context,
+    replaces a new head block that jumps to the end.
```
Comment (stuartarchibald) [inline numba/withcontexts.py]: Should this be `replaces the head block with a new block that jumps to the end`.?

File: numba/withcontexts.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        # Transform the block
+        _bypass_with_context(blocks, blk_start, blk_end, forwardvars)
+        _clear_blocks(blocks, body_blocks)
+
+
+bypass_context = _ByPassContextType()
+
+
+class _CallContextType(WithContext):
+    """A simple context-manager that tells the compiler lift the body of the
```
Comment (stuartarchibald) [inline numba/withcontexts.py]: s/compiler/compiler to/

**Comment by stuartarchibald** [comment]
Just done a bit of manual testing... some items for discussion

1. Seems like `ir.EnterWith` is not understood in object mode, or `@vectorize`, or `@stencil` and I imagine also `@guvectorize`.
2. Assuming it is desirable to fix 1. (which it may not be, in which case a more graceful error should be produced), the somewhat bizarre situation arises where object mode execution is potentially lifting out `with`-blocks as functions that are also running in object mode (cf. Loop lifting has same issue). I think this sort of thing might trigger it:
```python
def objmode_with_objmode():
  object()
  with call_context:
    object()
  return 1
```
3. Parfors is very broken. e.g.
```python
def lift():
  with call_context:
    pass
print(njit(parallel=True)(lift)())
```
gives:
```
  File "<path>/numba/numba/stencilparfor.py", line 49, in run
    if isinstance(call_list[0], StencilFunc):
IndexError: Failed at nopython (nopython frontend)
Failed at nopython (convert to parfors)
list index out of range
```
which was somewhat unexpected as there's no stencil anywhere near the source function!

Finally, is this to be documented in the developer documentation, or should that wait until more
features are added and the API etc stabilizes? Also, does this impact annotations, I don't think it will but am mindful that not breaking https://github.com/numba/numba/pull/2793 would be good!

File: numba/transforms.py [hunk 3]
```diff
@@ -299,3 +279,141 @@ def canonicalize_cfg(blocks):
...
+    # Can have any number of del
+    counters.pop(ir.Del, None)
+    # There MUST NOT be any other statements
+    if counters:
+        raise errors.CompilerError(
+            "illegal statements in with's head-block"
+            )
+
+
+def _cfg_nodes_in_region(cfg, region_begin, region_end):
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/transforms.py]: There are a number of things in `transforms.py` that can go into other modules.  I want to defer that task to later.

**Comment by sklam** [comment]
https://github.com/numba/numba/pull/3017#issuecomment-396592579 raised important concerns.  Given how this patch is not adding any user facing feature yet, I am ok to delay the merge until some of the concerns are resolved even if it will not make it this release cycle.  Particularly, the breaking of parfor features should be addressed first.

As for the objectmode support, there were no support for contextmanager in objectmode before.  It will fail in the bytecode analysis.  I think it is acceptable to skip objectmode support for this PR and just do a better errmsg for it.  Future patches should implement the objectmode support.

**Comment by ehsantn** [comment]
@sklam the Parfor-related issue was a simple bug that I just fixed.

**Comment by sklam** [comment]
Oh, looks like I need to resolve the `loc` review comments first.

File: numba/ir.py [hunk 4]
```diff
@@ -570,6 +570,31 @@ def list_vars(self):
...
+    """Enter a "with" context
+    """
+    def __init__(self, contextmanager, begin, end, loc):
+        """
+        Parameters
+        ----------
+        contextmanager : IR value
+        begin, end : int
+            The beginning and the ending offset of the with-body.
+        loc : int
```
Comment (stuartarchibald) [inline numba/ir.py]: `loc` is a `ir.Loc` instance?

File: numba/tests/test_withlifting.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        self.check_extracted_with(liftcall3, expect_count=2,
+                                  expected_stdout="A 1\nB 2\nC 47\n")
+        self.check_same_semantic(liftcall3)
+
+    def test_liftcall4(self):
+        with self.assertRaises(errors.TypingError) as raises:
+            njit(liftcall4)()
+        # Known error.  We only support one context manager per function
+        # for body that are lifted.
+        self.assertIn("re-entrant", str(raises.exception))
```
Comment (stuartarchibald) [inline numba/tests/test_withlifting.py]: Any view on doing something with non-existent context managers? I just accidentally forgot to import the `call_context`, but this equally applies to e.g.
```python
from numba import njit
def bar():
    with unknown_ctx:
        pass
njit(bar)()
```
which gives the same fairly spectacular traceback (as the one if you forget the import) ending with:
```
<path>/numba/numba/transforms.py in with_lifting(func_ir, typingctx, targetctx, flags, locals)
    293         _legalize_with_head(blocks[blk_start])
    294         cmkind = _get_with_contextmanager(func_ir, blocks, blk_start)
--> 295         sub = cmkind.mutate_with_body(func_ir, blocks, blk_start, blk_end,
    296                                       body_blocks, dispatcher_factory)
    297         sub_irs.append(sub)

AttributeError: Failed at nopython (nopython frontend)
'object' object has no attribute 'mutate_with_body'
```

**Comment by stuartarchibald** [comment]
Thanks for the fixes @sklam I made a couple more comments but am of the view that this can be merged whilst acknowledging that there is more work to be done/this is experimental and incremental development is expected.

File: numba/tests/test_withlifting.py (hunk 5 above)
Comment (sklam) [reply numba/tests/test_withlifting.py]: i think this should be fixed in this PR

File: numba/transforms.py [hunk 6]
```diff
@@ -315,11 +315,22 @@ def _get_with_contextmanager(func_ir, blocks, blk_start):
...
-            return dfn.value
+            ctxobj = dfn.value
+            if ctxobj is ir.UNDEFINED:
+                raise errors.CompilerError(
+                    "Undefined variable used as contextmanager",
+                    loc=blocks[blk_start].loc,
+                    )
+            if not hasattr(ctxobj, 'mutate_with_body'):
+                raise errors.CompilerError(
+                    "Unsupported use of contextmanager",
```
Comment (stuartarchibald) [inline numba/transforms.py]: Should that be `Unsupported context manager in use`, I think at this point the context manager statement has been identified and this is just duck type testing if it has an attr specific to Numba and is therefore supported?

**Comment by stuartarchibald** [comment]
Thanks for 995fdda @sklam, this can be merged as convenient.

**Comment by stuartarchibald** [comment]
@sklam tests are failing, think the error message update needs reflecting in the unit test catching the error. Thanks.

**Comment by sklam** [comment]
oops

**Comment by stuartarchibald** [comment]
Thanks for fixing. Merging.

**Pull request merged**